src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,231 @@
+  get<Tags::VPlus>(*weighted_char_fields_int) =
+      weight_char_field<Tags::VPlus>(v_plus_int, get<2>(char_speeds_int), -1.);
+  get<Tags::VMinus>(*weighted_char_fields_int) =
+      weight_char_field<Tags::VMinus>(v_minus_int, get<3>(char_speeds_int),
+                                      -1.);
+
+  get<Tags::VPsi>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VPsi>(v_psi_ext, get<0>(char_speeds_ext), 1.);
+  get<Tags::VZero<Dim>>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VZero<Dim>>(v_zero_ext, get<1>(char_speeds_ext),
+                                          1.);
+  get<Tags::VPlus>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VPlus>(v_plus_ext, get<2>(char_speeds_ext), 1.);
+  get<Tags::VMinus>(*weighted_char_fields_ext) =
+      weight_char_field<Tags::VMinus>(v_minus_ext, get<3>(char_speeds_ext), 1.);
+}
+}  // namespace CurvedScalarWave::BoundaryCorrections::CurvedScalarWave_detail
+
+namespace CurvedScalarWave::BoundaryCorrections {
+template <size_t Dim>
+UpwindPenalty<Dim>::UpwindPenalty(CkMigrateMessage* msg) noexcept
+    : BoundaryCorrection<Dim>(msg) {}
+
+template <size_t Dim>
+std::unique_ptr<BoundaryCorrection<Dim>> UpwindPenalty<Dim>::get_clone()
+    const noexcept {
+  return std::make_unique<UpwindPenalty>(*this);
+}
+
+template <size_t Dim>
+void UpwindPenalty<Dim>::pup(PUP::er& p) {
+  BoundaryCorrection<Dim>::pup(p);
+}
+
+template <size_t Dim>
+double UpwindPenalty<Dim>::dg_package_data(
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_psi,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_v_zero,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_plus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_v_minus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_gamma2,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_interface_unit_normal,
+    const gsl::not_null<tnsr::a<DataVector, 3, Frame::Inertial>*>
+        packaged_char_speeds,
+
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const Scalar<DataVector>& psi,
+
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& shift,
+    const tnsr::II<DataVector, Dim, Frame::Inertial>& inverse_spatial_metric,
+    const Scalar<DataVector>& constraint_gamma1,
+    const Scalar<DataVector>& constraint_gamma2,
+
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity)
+    const noexcept {
+  *packaged_gamma2 = constraint_gamma2;
+  *packaged_interface_unit_normal = interface_unit_normal;
+
+  {  // package characteristic fields
+    const auto char_fields =
+        characteristic_fields(constraint_gamma2, inverse_spatial_metric, psi,
```

Comment:
is there a `not_null` version of this function so you can avoid memory allocations? Anything that does memory allocations on the interfaces is rather expensive. The `char_fields` function also does 2 more memory allocations, so overall this part of the boundary correction is already at 4 (including the shift dot normal below) memory allocations per interface plus a bunch of copy of data. It's very little work compared to memory movement.

(see comment below for more thoughts :) )

Reply:
using return by reference of this function, AND removed the memory allocation in the `char_fields` function.